[stdlib] String index interchange, etc.

dabrahams opened the pull request.

Commit:
[stdlib] Index interchange, part I

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
Wow, some of those speedups are really surprising.

Commit:
[stdlib] String index interchange, part II (UTF16)

dabrahams commented:
@swift-ci Please smoke benchmark

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please smoke benchmark

dabrahams commented:
@swift-ci Please test source compatibility

palimondo commented:
All those "Hearts <3 ♥︎ 💘" and "Café 🍵"s will be lost in time, like tears in rain?

dabrahams commented:
@palimondo Gonna assume that's a joke that went over my head.  If it's not, please try to explain.

palimondo commented:
That was a [Blade Runner reference](https://en.wikipedia.org/wiki/Tears_in_rain_monologue)... and I was wondering if the removed documentation that was demonstrating use of `Index` with the "Hearts <3 ♥︎ 💘" and "Café 🍵"s would reappear in the new `String.Index` later, or was removed for good.

dabrahams commented:
Oh, removed for good because these types and therefore overloads are collapsing with others.  However I believe the same strings exist on the overloads with which they are collapsing, unless @natecook1000 was exceptionally, gratuitously creative ;-)

dabrahams commented:
I take it back; @natecook1000 was creative with good reason and many of them _do_ drop; He'll have to rewrite the docs for these things anyway.

dabrahams commented:
@swift-ci Please smoke benchmark

dabrahams commented:
@swift-ci Please smoke benchmark

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please clean smoke benchmark

Commit:
[stdlib] String index interchange, part III (UTF8)

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
Uh, right.  The linux failure is because I took away some SPI that Foundation was using for the Linux case only.  It's easy to bring back but I want to let the other tests finish.

Commit:
[stdlib] Bring back some SPI that Foundation-on-linux uses.

natecook1000 committed:
[stdlib] Minimal docs for the new string index

Commit:
[stdlib] UTF8View Iterator and count specialization

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented on stdlib/public/core/StringIndex.swift:
```diff
@@ -38,8 +39,11 @@ extension String.Index : Comparable {
   
-  public init(encodedOffset o: Int) {
-    _compoundOffset = UInt64(o << _Self._strideBits)
+  /// Creates a new index at the specified UTF-16 offset.
```
FYI, I aim to decouple this stuff from UTF-16 eventually.  That is really meant to be a code unit offset in some arbitrary encoding.  I'd have labeled it `utf16Offset` otherwise.

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

Commit:
[stdlib] Foundation SPI fixup

Commit:
[stdlib] Fix a test: String.UTF8View has its own Iterator now

Commit:
[stdlib] Small correction to UTF8 Iterator

This works either way I suppose, but the code is clearer now.

Commit:
[stdlib] UTF8View: count by iterating

The fancy code didn't turn out to be any faster.

Commit:
[stdlib] Speed String.UTF8View.count even more

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@moiseev Good point.  Maybe we can discuss this and figure out how to handle it.

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@moiseev Do you think [this failure](https://ci.swift.org/job/swift-PR-Linux/7824/consoleFull#852164225ee1a197b-acac-4b17-83cf-a53b95139a76) is related to that discussion?  I think it might be.

moiseev commented:
@dabrahams Not really. Regardless of what the non-failable init's type parameter is, the failable non-generic init wins, and also in Swift 3 mode the subscript returns a String by default... So... `as Substring` should be added somewhere, I guess.

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please clean test

dabrahams commented:
@swift-ci Please clean test

dabrahams commented:
@swift-ci Please test Linux Platform

dabrahams commented:
@moiseev I am unable to reproduce the failure locally, which is pretty strange.  I do see the init there, FWIW.

palimondo commented:
Isn't change in `StringFromLongWholeSubstring` from 15563 to 188ms (-98.8%), even in `Onone` a little bit suspect?

dabrahams commented:
@palimondo two answers:

1. The bug uncovered by [this failure|https://ci.swift.org/job/swift-PR-osx/7877/consoleFull#-1962751153fca400bf-2f4a-462e-b517-e058d770b2d7] would also result in some invalid speedups of SubString to String conversion... but not for the *whole* string case, which this is.

2. One of the optimizations in this branch changes those cases from doing ICU grapheme breaking to sort out individual Characters or can just suck in the bytes from the source string.  That is expected to be a *very* dramatic speedup, since ICU grapheme breaking is slow.

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test source compatibility

Commit:
[stdlib] Rebuild String.Index for UTF8View

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
Benchmarking again in light of #10153 which should have taken care of the MapReduceString regression.

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please benchmark

Commit:
[stdlib] Drop a couple bogus duplicate tests

Only the last of each set of dups gets run anyway.  See
https://bugs.swift.org/browse/SR-5362

Commit:
[stdlib] Extend String index conversion tests for interchange

test the new APIs as well as the legacy failing conversions.

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@airspeedswift Assuming tests pass on the 'bots, this is ready to go.  The changes to the tests verify that the semantics are as specified in the proposal.

dabrahams commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1096

@swift-ci Please benchmark

dabrahams commented:
Please test with following pull request:
https://github.com/apple/swift-corelibs-foundation/pull/1096

@swift-ci Please benchmark

dabrahams commented:
@swift-ci Please smoke test Linux platform

shahmishal commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

shahmishal commented:
@dabrahams Sorry GitHub did not deliver the webhook to ci.swift.org, its working now.

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please benchmark

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

shahmishal commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test

dabrahams commented:
@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

shahmishal commented:
@dabrahams Source compatibility suite does not support "Please test with following pull request:" feature. I will add the support and re-trigger the build.

shahmishal commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
The source compatibility failures are due to some change in Foundation, where `someData.reversed()` became ambiguous.

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test source compatibility

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test

dabrahams commented:
Please test with following pull request:
apple/swift-corelibs-foundation#1096

@swift-ci Please test
